lib/include/jxl/codestream_header.h

```diff
@@ -181,6 +190,11 @@ typedef struct {
    */
   JXL_BOOL have_animation;
 
+  /** Indicates if an intrinsic size is present in the codestream.
+   * Its dimensions are not included in the basic info.
+   */
+  JXL_BOOL have_intrinsic_size;
```

Comment:
Not sure if we really need that bool. We could also just always have intrinsic dimensions - when there are none signalled in the bitstream, decoder side they are just equal to the actual dimensions. Encoder side, we can have a special case where intrinsic dimensions of 0 x 0 (which is not representable in the jxl bitstream anyway) correspond to setting no intrinsic dimensions (so implicitly setting them equal to the actual dimensions).

Reply:
In any case, if we add the bool, it should be at the end, so we don't break ABI compatibility (not that it matters much, we'll break it anyway between the previous released version and the next one, but still).

Reply:
Moved it to the back.

Not having the bool might lead to some confusion: when the intrinsic dimensions are set exactly to the image dimension, one could interpret that as well as "not having intrinsic dimensions", but this might be precisely the intrinsic dimensions that one might want to signal sometimes. 
Going for signalling "no intrinsic dimension" with 0x0 would be fine with me, but I think it requires less explanation with that extra bool. Also it is consistent with the way we treat previews. 
What are the drawbacks of having that extra bool?

Added them to the end of the struct.

Reply:
Semantically, there is no difference between "not having intrinsic dimensions" and "having intrinsic dimensions equal to the image dimensions". If there are no intrinsic dimensions signaled, it has to be assumed that the intrinsic dimensions are equal to the actual dimensions, and if intrinsic dimensions are signaled that are equal to the actual dimensions, that doesn't mean anything else. It's like the difference between a JPEG without Exif orientation and a JPEG with Exif orientation set to the Identity.

There are no big drawbacks to having an extra bool, but it might be more convenient if you can just directly read `intrinsic_xsize`, `intrinsic_ysize` if you want to know the dimensions to render an image at, rather than having to do something like `if (bi.have_intrinsic_size) { render_width = bi.intrinsic_size.xsize; render_height = bi.intrinsic_size.ysize; } else { render_width = bi.xsize; render_height = bi.ysize; }`.

Previews are a bit different: not all images have a preview, and if there is none, then it doesn't make sense to talk about its dimensions.
In contrast, intrinsic size semantically always exists: if it's not explicitly signaled, then it defaults to the image dimensions, but that's just syntax (optimizing a few bytes by not having to signaling the trivial default) rather than a semantic difference.

So what I would propose is the following: no bool, and

- decoder: always sets intrinsic_xsize / intrinsic_ysize to their semantic values (equal to image dimensions if the jxl does not have explicitly signaled intrinsic dimensions)
- encoder: if intrinsic_xsize/ysize are set to zero or if they are set equal to the image dimensions, then don't explicitly signal intrinsic dimensions, otherwise set them to the given dimensions. I don't think we need to have a way to produce a jxl that explicitly (redundantly) signals the intrinsic dimensions to be equal to the image dimensions.

Reply:
ok, I have removed `have_intrinsic_size`. Not quite sure if this is really an improvement, because this way it is less consistent with the way it is handled internally. However I like the argument that it will be easier to use in the API, because `intrinsic_[xy]size` will always be something meaningful.